src/deploy/functions/validate.ts

```diff
@@ -49,3 +53,73 @@ export function functionIdsAreValid(functions: { id: string; platform: string }[
     throw new FirebaseError(msg);
   }
 }
+
+/**
+ * Validate secret environment variables setting, if any.
+ * A bad secret configuration can lead to a significant delay in function deploys.
+ *
+ * If validation fails for any secret config, throws a FirebaseError.
+ */
+export async function secretsAreValid(b: backend.Backend) {
+  const endpoints = backend
+    .allEndpoints(b)
+    .filter((e) => e.secretEnvironmentVariables && e.secretEnvironmentVariables.length > 0);
+  validatePlatformTargets(endpoints);
+  await validateSecretVersions(endpoints);
+}
+
+/**
+ * Ensures that all endpoints specifying secret environment variables target platform that supports the feature.
+ */
+function validatePlatformTargets(endpoints: backend.Endpoint[]) {
+  const supportedPlatforms = ["gcfv1"];
+  const unsupported = endpoints.filter((e) => !supportedPlatforms.includes(e.platform));
+  if (unsupported.length > 0) {
+    const errs = unsupported.map((e) => `${e.id}[platform=${e.platform}]`);
+    throw new FirebaseError(
+      `Tried to set secret environment variables on ${errs.join(", ")}. ` +
+        `Only ${supportedPlatforms.join(", ")} support secret environments.`
+    );
+  }
+}
+
+/**
+ * Validate each secret version referenced in target endpoints.
+ *
+ * A secret version is valid if:
+ *   1) It exists.
+ *   2) It's in state "enabled".
+ */
+async function validateSecretVersions(endpoints: backend.Endpoint[]) {
+  const validate = async (s: backend.SecretEnvVar) => {
+    const sv = await getSecretVersion(s.projectId, s.secret, "latest");
+    if (s.version == null) {
+      logLabeledSuccess(
+        "functions",
+        `resolved secret version of ${clc.bold(s.secret)} to ${clc.bold(sv.version)}.`
+      );
+      s.version = sv.version;
+    }
+    if (sv.state !== "ENABLED") {
+      throw new FirebaseError(
+        `Expected secret ${s.secret}@${s.version} to be in state ENABLED not ${sv.state}.`
+      );
+    }
+  };
+
+  const validations: Promise<void>[] = [];
+  for (const e of endpoints) {
+    for (const s of e.secretEnvironmentVariables! || []) {
+      validations.push(validate(s));
+    }
+  }
+  const results = await utils.allSettled(validations);
+
+  const errs: { message: string }[] = results
+    .filter((r) => r.status === "rejected")
+    .map((r) => (r as utils.PromiseRejectedResult).reason as { message: string });
+  if (errs.length) {
+    const msg = errs.map((e) => e.message).join(", ");
+    throw new FirebaseError(`Failed to validate secret versions: ${msg}`);
```

Comment:
FYI: FirebaseError also has a `children` option. Not sure I'm a fan of it (since it seems to only show up in the debug log) so I'm totally cool with putting the message here too.

Reply:
Great suggestion! Looks pretty good:

```
✔  functions: resolved secret version of MY_SECRET to 3.
Error: Failed to validate secret versions:
- FirebaseError HTTP Error: 404, Secret [projects/829294825753/secrets/MY_SECRET123] not found or has no versions.
- FirebaseError Expected secret MY_SECRET@3 to be in state ENABLED not DESTROYED.
```